kcidb/tools/kcidb_match.py

```diff
@@ -0,0 +1,429 @@
+            log_excerpt = kcidb_io_object[key][0].get("log_excerpt", "")
+            log_url = kcidb_io_object[key][0].get("log_url", "")
+            return search_regex(log_regex, log_excerpt, log_url)
+        return False
+
+
+class IncidentGenerator:
+    """Class to generate incidents"""
+    def __init__(self, db_name=DB_NAME):
+        self.db = PatternDatabase(db_name)
+        self.validator = PatternValidator()
+
+    def create_incident(self, kcidb_io_object, issue_id, issue_version):
+        """Create and return an incident object"""
+        if tests := kcidb_io_object.get('tests'):
+            type_id_key = "test_id"
+            type_id_value = tests[0]['id']
+        elif builds := kcidb_io_object.get('builds'):
+            type_id_key = "build_id"
+            type_id_value = builds[0]['id']
+        else:
+            raise ValueError("The KCIDB IO object must contain at least "
+                             "one non-empty test or build")
+
+        unique_string = f"{issue_id}{issue_version}{type_id_value}"
+        incident_id = f"{ORIGIN}:" \
+                      f"{hashlib.sha256(unique_string.encode()).hexdigest()}"
+
+        return {
+            'id': incident_id,
+            'origin': ORIGIN,
+            'issue_id': issue_id,
+            'issue_version': issue_version,
+            'present': True,
+            type_id_key: type_id_value,
+        }
+
+    def generate_incident_on_match(self, kcidb_io_object, issue_id,
+                                   issue_version, issue_pattern_object):
+        """Generate incident if issue pattern is found in a build/test
+        object"""
+        incident = {}
+
+        if self.validator.match_fields(issue_pattern_object, kcidb_io_object):
+            incident = self.create_incident(kcidb_io_object, issue_id,
+                                            issue_version)
+
+        return incident
+
+    def generate_incidents_from_db(self, kcidb_io_object):
+        """Generate incidents by trying to match the kcidb_io_object
+        against the patterns saved in the database"""
+        incidents = []
+
+        for row in self.db.get_all_patterns():
+            issue_id, issue_version, pattern_object_json = row
+            pattern_object = json.loads(pattern_object_json)
+            incident = self.generate_incident_on_match(
+                kcidb_io_object, issue_id, issue_version, pattern_object)
+            if incident:
+                incidents.append(incident)
+
+        return {
+            "version": KCIDB_IO_VERSION,
+            "incidents": incidents
+        }
+
+    def generate_incidents_from_test(self, test):
+        """Generate incident from test object"""
+        kcidb_io_object = {"tests": [test._data],
+                           "builds": [test.build._data],
+                           "checkouts": [test.build.checkout._data]}
+        return self.generate_incidents_from_db(kcidb_io_object)
```

Comment:
You cannot make an I/O object from an OO object in general, even if it looks like you can. You just cannot count on that. An OO object is a processed I/O object, and there can be data loss. Instead deal with OO objects directly everywhere. Where you need to process I/O objects from stdin and command-line interface, load them into an sqlite database using the database client, get them as OO objects from there, and then process.

OO objects were specifically made to make the things you're doing here easier to do. Like walk related objects and so on. We can have a call this week and go over all the concerns and options regarding this.

Reply:
BTW, there's also a schema for the raw OO data in `kcidb.orm.data`.